core/src/HIP/Kokkos_HIP_KernelLaunch.hpp

```diff
@@ -520,9 +520,7 @@ struct HIPParallelLaunchKernelInvoker<DriverType, LaunchBounds,
                             HIPInternal const *hip_instance) {
     const auto hip_device = hip_instance->m_hipDev;
     // Wait until the previous kernel that uses the constant buffer is done
-    std::lock_guard<std::mutex> lock(HIPInternal::constantMemMutex[hip_device]);
-    KOKKOS_IMPL_HIP_SAFE_CALL(
-        hipEventSynchronize(HIPInternal::constantMemReusable[hip_device]));
+    const auto lock = HIPInternal::constantMemReusable[hip_device].acquire();
```

Comment:
This sets `ready` to `false`. It might be premature because the copies and the launch might still fail.